src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/PropagateCancellationTokensWhenPossible.cs

```diff
@@ -33,11 +39,151 @@ public abstract class PropagateCancellationTokensWhenPossibleAnalyzer : Diagnost
+                // CancellationToken value. However, check for null to ensure we don't crash on invalid code.
+                if (argumentValue == null)
+                {
+                    // The code is invalid. The compiler error will be enough warning for the developer.
+                    return false;
+                }
+
+                if (!IsCancellationTokenNone(argumentValue, cancellationTokenType))
+                {
+                    // A non-default CancellationToken is being passed.
+                    return false;
+                }
+            }
+
+            // Check that a suitable variable of type CancellationToken is in scope.
+            if (!IsCancellationTokenVariableInScope(invocation, cancellationTokenType, compilation))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private static bool IsCancellationTokenNone(IOperation value, INamedTypeSymbol cancellationTokenType)
+        {
+            // This method returns true for CancellationToken.None, explicitly-passed default(CancellationToken),
+            // and compiler-generated default(CancellationToken) (e.g. as the value of an optional parameter).
+            // Note that default(CancellationToken) is equivalent to CancellationToken.None.
+
+            Debug.Assert(value.Type == cancellationTokenType);
+
+            switch (value.Kind)
+            {
+                case OperationKind.DefaultValue:
+                    return true;
+                case OperationKind.PropertyReference:
+                    var property = ((IPropertyReferenceOperation)value).Property;
+                    if (property.Name == "None" && property.ContainingType == cancellationTokenType)
+                    {
+                        return true;
+                    }
+                    break;
+            }
+
+            return false;
+        }
+
+        private static bool IsCancellationTokenVariableInScope(IInvocationOperation invocation, INamedTypeSymbol cancellationTokenType, Compilation compilation)
+        {
+            var semanticModel = compilation.GetSemanticModel(invocation.Syntax.SyntaxTree);
+            int position = invocation.Syntax.SpanStart;
+            return semanticModel.LookupSymbols(position).Any(IsUsableCancellationTokenVariable);
+
+            bool IsUsableCancellationTokenVariable(ISymbol symbol)
+            {
+                switch (symbol.Kind)
+                {
+                    case SymbolKind.Field:
+                        var field = (IFieldSymbol)symbol;
+                        return field.Type == cancellationTokenType;
+                    case SymbolKind.Local:
+                        var local = (ILocalSymbol)symbol;
+                        return local.Type == cancellationTokenType && !local.IsInaccessibleLocal(position);
```

Comment:
>local.IsInaccessibleLocal(position) [](start = 71, length = 35)

This isn't going to be enough, because of uninitialized local variables. You have a cancellationtoken that's in scope, but hasn't been set up yet.